### DISCUSSION TITLE
The magic 299 or can we go lower?

### DISCUSSION THREAD

**Discussion by AnAnalogGuy** [gcormier/megadesk#167] [General]
Hi Greg,

Looking at the code, i found DANGER_MIN_HEIGHT (162 + HYSTERESIS) 
where HYSTERESIS == 137 (minimum raw step width)
https://github.com/gcormier/megadesk/blob/eada425798dad95a4d8fdc1375178ac9d3189996/code/src/megadesk.cpp#L143

with your notes:
// Changing these might be a really bad idea. They are sourced from
// decoding the OEM controller limits.

That leads to  299 (162 + 137 = 299) which is the minimum position the table can be moved to with megadesk. 

However, when calibrating my desk, it reaches raw 96. Now, for my "low rider sitting position" i'd like to get those extra approx 2cm lower, if possible. 

@gcormier from your experience, do you think it will be a problem to override the DANGER_MIN_HEIGHT? I'm just asking myself what could go wrong? My thoughts are:

- With the calibration functionality the desk already moves into this position
- These extra 299 minus 96 i consider a security margin since products from mass production might have fluctuations. So for some the extra margin is needed, for some not. 
- Isn't there a protection in place anyway? I mean, doesn't the calibration function demonstrate that there obv is some sort of stop mechanism reacting on increasing power consumption (since when physical resistance occurs, the motor is staring to pull more power which then obv. triggers an electrical "shutdown")?

So if i would set my minimum height to something above 96, say 110, it still shoud have a small margin to not trigger that protection. And in case something should go wrong, it would only be once since i can change the setup again if i see it's going to be a problem. 

What do you think @gcormier, is there something I'missing?

Thanks a lot!

**Comment by gcormier** [comment]
You need to account for the "coasting" or deceleration when you let go of the button. 

But otherwise, in theory you could probably go lower. Two factors contributed to this,

1. The calibration function was not known when I did all the reverse engineering, nor was it implemented on megadesk until much later. I agree now, it is likely the motors should be able to protect themselves.
2. There was no way I was going to ship the pre-flashes version with something that had an element of risk for damaging the desk, hence the generous buffer.


If you decide to go for it, please report back :)

**Reply by AnAnalogGuy** [reply]
I'm fully with you and would have done the same when replacing the original pcb - no way i would have made it behave else than the original. I was just wondering if you might have had any other reasons or insights before i will give it a try. And sure, i will report back - will take a while as i first have to order a suitable AVR programmer.

**Reply by AnAnalogGuy** [reply]
Btw, one more reason to include some buffer is both legs not always having the same remaining "length" (which also is something that gets adjusted during recalibration).

**Comment by AnAnalogGuy** [comment]
Did i get it right and [env:attiny841_serial] would be the right "action" to choose for megadesk? Anything special to keep an eye on?

Has anyone used VS Code on Windows with PlatformIO to build and upload the hex to the microcontroller? It seems PlatformIO uses avrdude. Would be a very integrated and straight forward push-one-button action if it really works ootb.

**Reply by gcormier** [reply]
Yes, that's the right action. vscode + pio is what I use for all my development.

**Comment by AnAnalogGuy** [comment]
@gcormier 
I'm just thinking about how to implement the lower `DANGER_MIN_HEIGHT`. Instead of hardcoding the value, I'm thinking if i should use the recalibration method, to set `DANGER_MIN_HEIGHT` to the position after the recal plus a buffer. Like this, the `DANGER_MIN_HEIGHT` gets validated by the desks calibration functionality which should never return an invalid value, no matter individual differences per desk or deviations over time. I could (try to :-)) add a build flag to make this "feature" controllable at compile time.

---
After going a bit deeper into the code, I guess the recal idea won't work as if I correctly understand, input processing is done single threaded in sequential order within `loop()`, so during the recal procedure `currentHeight `won't be updated and i won't be table to get the actual value before the next iteration of `loop()`

**Comment by AnAnalogGuy** [comment]
One more thing i wasn't able to figure out: Why is it that SCK on the megadesk board is used as TX channel in the serial communication? I'm just wondering as i would expect it's a different signal than a clock and the same port, if i get it right, is actually used as clock when connected to a an eeprom writer to do the firmware update. Or did i get this wrong?

![grafik](https://github.com/gcormier/megadesk/assets/70239317/b29f4f6f-5c20-444a-aef0-3ec2e885694e)

![grafik](https://github.com/gcormier/megadesk/assets/70239317/c7419c54-7611-45ac-bf99-fe75cb617d72)

![grafik](https://github.com/gcormier/megadesk/assets/70239317/db2fb2fa-2ae5-4ddf-a521-ed55fbe6c33b)

**Reply by gcormier** [reply]
SCK is also RXD1 which is a hardware UART on the 841.

**Comment by AnAnalogGuy** [comment]
While trying to fully understand the code i came across this section:

![grafik](https://github.com/gcormier/megadesk/assets/70239317/eeb4f840-440f-490c-a752-811bb8436f32)

https://github.com/gcormier/megadesk/blob/eada425798dad95a4d8fdc1375178ac9d3189996/code/src/megadesk.cpp#L775

Shouldn't `currentHeight `in the red squares be `targetHeight`? Since, if i get it right, the goal of the second condition is to avoid driving the motors over the upper and lower limits? Checking against currentHeight would not always avoid this. 

There is a second check for the limits in a different part of the code, which might be the reason why this never led to a problem.

**Reply by gcormier** [reply]
I believe the intention is that if the desk is already at one of the limits, it will avoid starting any movement commands altogether.

**Comment by AnAnalogGuy** [comment]
One more question if you a couple of minutes: 

![grafik](https://github.com/gcormier/megadesk/assets/70239317/f05c7fb6-47a2-46b3-8f8c-0fa9ba2cc2c1)

Did i get this right:

At the begining of every iteration of loop() the current desk position is read from the motors and `currentHeight `is updated. After this, commands (serial commands or manual buttons) are evaluated. If recal is triggered, the control flow is:

- In State `STARTING_RECAL`, set new State `RECAL`, break
- next iteration of loop, starting with evaluating current height
- In State `RECAL`, set new State `END_RECAL`, break
- next iteration of loop, starting with evaluating current height
- In State `END_RECAL`

So, when the program reaches `END_RECAL`, `currentHeight `has already been updated to the most recent height after calibration has been done (moved the table)? So i could automatically set `DANGER_MIN_HEIGHT `to the minimun heigth after calibration?

![grafik](https://github.com/gcormier/megadesk/assets/70239317/d904556d-7096-4e54-80fe-27636b8d8114)

Could i use eeprom slot 14 to persist DANGER_MIN_HEIGHT (and read it back after a restart, as it's not a constant any more)?
![grafik](https://github.com/gcormier/megadesk/assets/70239317/891cf1fe-ce87-44ba-84aa-f4648e0705ba)

**Reply by gcormier** [reply]
Yes you could use slot 14. And yes if you want you could just set the danger min height to the value after calibration is finished.

**Reply by henrikmathisen** [reply]
> One more question if you a couple of minutes:
> 
> ![grafik](https://private-user-images.githubusercontent.com/70239317/343137288-f05c7fb6-47a2-46b3-8f8c-0fa9ba2cc2c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4MjgxNjcsIm5iZiI6MTczMzgyNzg2NywicGF0aCI6Ii83MDIzOTMxNy8zNDMxMzcyODgtZjA1YzdmYjYtNDdhMi00NmIzLThmOGMtMGZhOWJhMmNjMmMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDEwNTEwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2YzRlZWIwY2UzYTIwNzc2OTk1MjY4OTI2YTNiMThjNWY4MDc2YWZhMDQ3ZWEyMzZjMjViZDBkMmIyM2NiZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.QFsnfSUyt0sslc0Pb1w0zCtbRK8wKElqNZOnmOl2WGc)
> 
> Did i get this right:
> 
> At the begining of every iteration of loop() the current desk position is read from the motors and `currentHeight `is updated. After this, commands (serial commands or manual buttons) are evaluated. If recal is triggered, the control flow is:
> 
>     * In State `STARTING_RECAL`, set new State `RECAL`, break
> 
>     * next iteration of loop, starting with evaluating current height
> 
>     * In State `RECAL`, set new State `END_RECAL`, break
> 
>     * next iteration of loop, starting with evaluating current height
> 
>     * In State `END_RECAL`
> 
> 
> So, when the program reaches `END_RECAL`, `currentHeight `has already been updated to the most recent height after calibration has been done (moved the table)? So i could automatically set `DANGER_MIN_HEIGHT `to the minimun heigth after calibration?
> 
> ![grafik](https://private-user-images.githubusercontent.com/70239317/343143396-d904556d-7096-4e54-80fe-27636b8d8114.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4MjgxNjcsIm5iZiI6MTczMzgyNzg2NywicGF0aCI6Ii83MDIzOTMxNy8zNDMxNDMzOTYtZDkwNDU1NmQtNzA5Ni00ZTU0LTgwZmUtMjc2MzZiOGQ4MTE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDEwNTEwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzYzY5YzdkMmIzNjllMGE3NTdkNGIwY2M0ZDQ4ZjQxMTg2NjZlZjkzOWRlNzY2M2UwNGVkZmQzYmZlZDYzZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8fqrDwQre7iITTTo-ff70p8XHqoevCepwJ-oSCR2Bgc)
> 
> Could i use eeprom slot 14 to persist DANGER_MIN_HEIGHT (and read it back after a restart, as it's not a constant any more)? ![grafik](https://private-user-images.githubusercontent.com/70239317/343146956-891cf1fe-ce87-44ba-84aa-f4648e0705ba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4MjgxNjcsIm5iZiI6MTczMzgyNzg2NywicGF0aCI6Ii83MDIzOTMxNy8zNDMxNDY5NTYtODkxY2YxZmUtY2U4Ny00NGJhLTg0YWEtZjQ2NDhlMDcwNWJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDEwNTEwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjZmQxMGJiZjEwODg4OTBhMGIwODc4NDE3MWIwMTg5ZjlkMjlhY2IxMzlkZTRjMTBiNDllYmZhYTMzZTUwOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VkJzTPmBShFIz_FprYIBTni2YI6u_JeOU1MaoZsJjCg)

Did you try updating `DANGER_MIN_HEIGHT` after calibration? Do you still have the code? :)

**Reply by AnAnalogGuy** [reply]
Yes, I still have the code, but I didn't had time to test it, so I haven't pushed it to the repo yet.

**Reply by henrikmathisen** [reply]
> Yes, I still have the code, but I didn't had time to test it, so I haven't pushed it to the repo yet.

Alright, I hope you let us know when you do, I tried tinker with it but I'm not familiar enough with cpp to get it to work. <3

**Reply by AnAnalogGuy** [reply]
You are aware that we are talking about approx 1 cm?
Do a "recalibration" to get an idea where the position will be at.
